mlkem/poly.c

```diff
@@ -220,36 +220,62 @@ void poly_compress(uint8_t r[KYBER_POLYCOMPRESSEDBYTES], const poly *a)
     {
         // REF-CHANGE: Hoist scalar decompression into separate function
-        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[0] >> 0) & 0xF);
-        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[0] >> 4) & 0xF);
-        a += 1;
+        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[i] >> 0) & 0xF);
+        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[i] >> 4) & 0xF);
     }
+
+
+
     #elif (KYBER_POLYCOMPRESSEDBYTES == 160)
-    unsigned int j;
-    uint8_t t[8];
-    for (i = 0; i < KYBER_N / 8; i++)
+    const int num_blocks = KYBER_N / 8;
+    for (int i = 0; i < num_blocks; i++)
+        __CPROVER_assigns(i, __CPROVER_object_whole(r))
+        __CPROVER_loop_invariant(i >= 0)
+        __CPROVER_loop_invariant(i <= num_blocks)
+        __CPROVER_loop_invariant(num_blocks == 32)
+        __CPROVER_loop_invariant(__CPROVER_forall { int k; k >= 0 && k <= (8 * i - 1) ==>
+                                 ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q )})
+        __CPROVER_decreases(num_blocks - i)
     {
+        uint8_t t[8];
+        const int offset = i * 5;
         // REF-CHANGE: Explicitly truncate to avoid warning about
-        // implicit truncation in CBMC.
-        t[0] = 0x1F &  (a[0] >> 0);
-        t[1] = 0x1F & ((a[0] >> 5) | (a[1] << 3));
-        t[2] = 0x1F &  (a[1] >> 2);
-        t[3] = 0x1F & ((a[1] >> 7) | (a[2] << 1));
-        t[4] = 0x1F & ((a[2] >> 4) | (a[3] << 4));
-        t[5] = 0x1F &  (a[3] >> 1);
-        t[6] = 0x1F & ((a[3] >> 6) | (a[4] << 2));
-        t[7] = 0x1F &  (a[4] >> 3);
-        a += 5;
-
-        for (j = 0; j < 8; j++)
+        // implicit truncation in CBMC and unwind loop for ease
+        // of proof.
+
+        // Decompress 5 8-bit bytes (so 40 bits) into
+        // 8 5-bit values stored in t[]
+        t[0] = 0x1F &  (a[offset + 0] >> 0);
+        t[1] = 0x1F & ((a[offset + 0] >> 5) | (a[offset + 1] << 3));
+        t[2] = 0x1F &  (a[offset + 1] >> 2);
+        t[3] = 0x1F & ((a[offset + 1] >> 7) | (a[offset + 2] << 1));
+        t[4] = 0x1F & ((a[offset + 2] >> 4) | (a[offset + 3] << 4));
+        t[5] = 0x1F &  (a[offset + 3] >> 1);
+        t[6] = 0x1F & ((a[offset + 3] >> 6) | (a[offset + 4] << 2));
+        t[7] = 0x1F &  (a[offset + 4] >> 3);
+
+        // and copy to the correct slice in r[]
+        for (int j = 0; j < 8; j++)
+            __CPROVER_assigns(j, __CPROVER_object_whole(r))
+            __CPROVER_loop_invariant(j >= 0)
+            __CPROVER_loop_invariant(j <= 8)
+            __CPROVER_loop_invariant(i >= 0)
+            __CPROVER_loop_invariant(i <= num_blocks)
+            __CPROVER_loop_invariant(__CPROVER_forall { int k; k >= 0 && k <= (8 * i + j - 1) ==>
+                                     ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q )})
+            __CPROVER_decreases(8 - j)
         {
-            // REF-CHANGE: Truncation happened before
             r->coeffs[8 * i + j] = ((uint32_t) t[j] * KYBER_Q + 16) >> 5;
```

Comment:
It's rather ugly that for the `(KYBER_POLYCOMPRESSEDBYTES == 128)` case this computation is moved to a separate function, while for the ` (KYBER_POLYCOMPRESSEDBYTES == 160)` case, it's still inlined. 

I don't understand why it needs to be moved into a separate function at all. Can we keep it inlined for both?

Reply:
Good catch, it should be one or the other. I vaguely recall that proof development was simpler when hoisting out the scalar functions, but apparently everything still works when inlining their bodies, as witnessed by #165 -- so either way will be OK.

For this PR, however, I'd stick with the present approach of _having_ wrappers for scalar functions, and use them consistently. I added a commit using `scalar_decompress_

Reply:
Okay, I'm fine with consistently using the scalar functions for now. Thanks.